src/engine/BMInterfaceTournament.php

```diff
@@ -0,0 +1,987 @@
+
+            if (!$this->validate_leave_tournament($userId, $tournId, $tournament)) {
+                return NULL;
+            }
+
+            $success = $this->remove_user_from_tournament($userId, $tournId, $tournament->nPlayers);
+
+            if (!$success) {
+                // something's gone wrong between validation and attempting to remove the user
+                $tournament = $this->load_tournament($tournId);
+                $validation = $this->validate_leave_tournament($userId, $tournId, $tournament);
+
+                if ($validation) {
+                    error_log('Paradoxical validation success in BMInterface::leave_tournament');
+                    $this->set_message('Tournament leave failed even though it should have succeeded');
+                }
+                return NULL;
+            }
+
+            $this->unwatch_tournament($userId, $tournId);
+        } catch (BMExceptionDatabase $e) {
+            $this->set_message('Cannot leave tournament because a player or tourn ID was not valid');
+            return NULL;
+        } catch (Exception $e) {
+            error_log(
+                'Caught exception in BMInterface::leave_tournament: ' .
+                $e->getMessage()
+            );
+            $this->set_message("Tournament leave failed: $e");
+            return NULL;
+        }
+
+        $this->update_tournament($tournId);
+        $this->set_message('You have successfully left this tournament');
+        return TRUE;
+    }
+
+    /**
+     * Validate whether a player can leave a tournament
+     *
+     * @param int $userId
+     * @param int $tournId
+     * @param BMTournament $tournament
+     * @return bool
+     */
+    protected function validate_leave_tournament($userId, $tournId, $tournament) {
+        if (is_null($tournament)) {
+            $this->set_message("Tournament $tournId does not exist");
+            return NULL;
+        }
+
+        if ($tournament->tournamentState > BMTournamentState::JOIN_TOURNAMENT) {
+            $this->set_message('The tournament has already started.');
+            return NULL;
+        }
+
+        if (!in_array($userId, $tournament->playerIdArray)) {
+            $this->set_message('You are not part of this tournament.');
+            return NULL;
+        }
+
+        if (count($tournament->playerIdArray) == $tournament->nPlayers) {
+            $this->set_message('You cannot leave when the tournament is full.');
+            return NULL;
+        }
+
+        return TRUE;
+    }
+
+    /**
+     * Attempt to remove a user from a tournament in the database
+     *
+     * @param int $userId
+     * @param int $tournId
+     * @param int $nPlayers
+     * @return bool
+     */
+    protected function remove_user_from_tournament($userId, $tournId, $nPlayers) {
+        // query is written in this way to avoid possible race conflicts between
+        // multiple people trying to act on a tournament at the same time
```

Comment:
Are we not loading and then saving the tournament in a transaction?  It makes me nervous that we're planning to handle races in code; we probably won't be good at it.

Reply:
This is now obsolete, because we noticed that there were issues with potential nested transactions.